### PR TITLE
Fixed host issue for kafka-interactive-queries

### PR DESCRIPTION
**Pull request by nehabhardwaj01**
Currently, any string provided as the APPLICATION_HOST will start the REST Service on localhost (0.0.0.0).

This fix will enhance the functionality by- 
- Validating the HOST and raising the exception for any invalid/ unreachable/ unavailable host.
- Starts the REST service on the provided host and port (If valid). 

Issue link: https://github.com/confluentinc/kafka-streams-examples/issues/127

### CODE REVIEW

**Comment by nehabhardwaj01** [comment]
[clabot:check]

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 1]
```diff
@@ -236,6 +239,145 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
     assertTrue(keyValueBean.getKey().startsWith("streams"));
     assertThat(keyValueBean.getValue(), equalTo(3L));
   }
+  
+  @Test
+  public void shouldStartRestApiOnAnyValidHost() throws Exception {
+    // Race condition caveat:  This two-step approach of finding a free port but not immediately
```
Comment (guozhangwang) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Hmm.. this is a it worrisome to me, if we cannot have stable examples repo unit test. Note that we have regular cron jobs to run the unit tests and if anything fails we need to look into it. Having an unstable test due to port binding would reduce the sensitivity of the test. Could we improve on fixing it?

Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Hi @guozhangwang, The original tests in the class also use the same way of binding and the comment is mentioned for them as well.
Although this comment serves as a corner case, still would like to know any suggestions for improvement.

Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: If we can wrap the search and bind in a retry loop, I think we'd be all good. It's what I've done in the past.

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 2]
```diff
@@ -67,9 +66,71 @@
   private KafkaStreams streams;
   private MusicPlaysRestService restProxy;
   private int appServerPort;
+  private static final List<Song> songs = Arrays.asList(new Song(1L,
```
Comment (bbejeck) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: These sample songs are loaded from a file (thanks @ybyzek ) in `io.confluent.examples.streams.interactivequeries.kafkamusic.KafkaMusicExampleDriver`.  Maybe we could re-use the same approach here?

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java (hunk 1 above)
Comment (bbejeck) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: This warning comment is pre-existing.  

As for an approach that would help with failures due to the race condition mentioned, the only thing that comes to mind ATM is to do a try/catch around the block of code setting the port and starting the rest proxy server with some reasonable number of retries (2-3?).

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 3]
```diff
@@ -236,6 +239,145 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
     assertTrue(keyValueBean.getKey().startsWith("streams"));
     assertThat(keyValueBean.getValue(), equalTo(3L));
   }
+  
+  @Test
+  public void shouldStartRestApiOnAnyValidHost() throws Exception {
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: If all this test is verifying is that all the right binding and metadata advertisement happens, I think it would be better to skip the actual IQ logic test (the exiting test already checks it). Similar to your other test that verifies an exception binding to an invalid host, we could just do the setup and do a quick check that everything is hooked up properly.

Then again, I'm not very familiar with this example. Is this suggestion valid?

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 4]
```diff
@@ -69,21 +69,40 @@
...
+    // Produce sample data to the input topic before the tests starts.
+    Properties producerConfig = new Properties();
+    producerConfig.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+    producerConfig.put(ProducerConfig.ACKS_CONFIG, "all");
+    producerConfig.put(ProducerConfig.RETRIES_CONFIG, 0);
+    producerConfig.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+    producerConfig.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+    IntegrationTestUtils
+            .produceValuesSynchronously(WordCountInteractiveQueriesExample.TEXT_LINES_TOPIC, inputValues,
+                    producerConfig);
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: I think it would be better to keep this in the method(s) that need(s) it. The reason is that not all tests needs to consume this data, but it will get populated before every test. It's logically fine, but it does cause our tests to run slower.

In fact, I'd recommend leaving this method as you have it, and just removing the `@BeforeClass` annotation and directly calling this method at the start of the relevant tests.

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 5]
```diff
@@ -67,9 +66,71 @@
...
+                                                        new Song(12L,
+                                                                 "The Calling",
+                                                                 "Hilltop Hoods",
+                                                                 "The Calling",
+                                                                 "Hip Hop")
+                                                        );
 
   @BeforeClass
-  public static void createTopics() {
+  public static void createTopicsAndProduceDataToInputTopics() {
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: Please apply me feedback from the `WordCountInteractiveQueriesExampleTest`.

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 6]
```diff
@@ -210,17 +188,91 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+                                              CLUSTER.schemaRegistryUrl(),
+                                              appServerPort,
+                                              TestUtils.tempDirectory().getPath(),
+                                              host);
+    restProxy = KafkaMusicExample.startRestProxy(streams, new HostInfo(host, appServerPort));
+  }
+
+  @After
+  public void shutdown() throws Exception {
+    restProxy.stop();
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: should we catch and log an exception here, so we'll still try to close `streams` if `restProxy.stop` fails?

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 7]
```diff
@@ -210,17 +188,91 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+                              songCountPlayBean(songs.get(1), 5L),
+                              songCountPlayBean(songs.get(7), 5L),
+                              songCountPlayBean(songs.get(2), 4L)
+                              )
+                );
+
+  }
+
+  @Test
+  public void shouldDemonstrateInteractiveQueriesOnAnyValidHost() throws Exception {
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: Similar to my feedback from the `WordCountInteractiveQueriesExampleTest`, do we need the full validation of IQ here, or is it sufficient to just verify the right host got bound?

**Comment by vvcephei** [comment]
Hey @nehabhardwaj01 , 

I left a few comments/suggestions/nitpicks, but it looks good overall.

Thanks for the PR!
-John

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java (hunk 3 above)
Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: @vvcephei Apart from verifying binding and metadata advertisement, We are testing all those scenarios(IQ logic test) just to ensure that all rest endpoints works as expected with this host as well.
If anyhow, they seem irrelevant, we can surely make the suggested changes.

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java (hunk 4 above)
Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: @vvcephei, Maybe there's some confusion, Actually, The @BeforeClass annotation ensures that any *computationally expensive setup is executed only once* for all those tests that share this setup.
However, @Before will run before each test case.

Reference: http://junit.sourceforge.net/javadoc/org/junit/BeforeClass.html

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java (hunk 6 above)
Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: Sure, @vvcephei It'll be a better approach. Will apply it.

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java (hunk 2 above)
Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: Sure, @bbejeck The approach seems good and applicable. Will apply them. Thanks for pointing it out.

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java (hunk 1 above)
Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: @bbejeck @vvcephei, The PR contains a try/catch block for binding host and port already.
My bad, missed a test case for that, Will write a test case to elaborate its use.

Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Agree, Adding a try/catch block will add more robustness to the tests.
Will do the change. :+1:

**Comment by nehabhardwaj01** [comment]
Hi @vvcephei, 

Thanks for reviewing the PR, I've answered few of your queries and raised questions for those I might need more clarity.
I've committed the changes and will be great if you can review them once.

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java (hunk 4 above)
Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Ah, of course. I misread the code. My apologies.

**Comment by guozhangwang** [comment]
made another pass and do not have further comments. @vvcephei @bbejeck please feel free to commit / pint merge if it lgty.

File: src/main/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExample.java [hunk 8]
```diff
@@ -202,9 +202,10 @@ public static void main(String[] args) throws Exception {
   }
 
 
-  static WordCountInteractiveQueriesRestService startRestProxy(final KafkaStreams streams, final int port)
+  static WordCountInteractiveQueriesRestService startRestProxy(final KafkaStreams streams, final int port, final
+                                                               String host)
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExample.java]: nit: fix formatting
```
startRestProxy(final KafkaStreams streams,
               final int port,
               final String host)
```

File: src/main/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesRestService.java [hunk 9]
```diff
@@ -260,8 +264,20 @@ void start(final int port) throws Exception {
...
+    ServerConnector connector = new ServerConnector(jettyServer);
+    connector.setHost(hostInfo.host());
+    connector.setPort(port);
+    jettyServer.addConnector(connector);
+  
+    context.start();
+    
+    try {
+      jettyServer.start();
+    } catch (java.net.SocketException exception) {
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesRestService.java]: nit: add `final`

File: src/main/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesRestService.java [hunk 10]
```diff
@@ -260,8 +264,20 @@ void start(final int port) throws Exception {
     ServletContainer sc = new ServletContainer(rc);
     ServletHolder holder = new ServletHolder(sc);
     context.addServlet(holder, "/*");
-
-    jettyServer.start();
+    
+    ServerConnector connector = new ServerConnector(jettyServer);
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesRestService.java]: nit: add `final`

File: src/main/java/io/confluent/examples/streams/interactivequeries/kafkamusic/MusicPlaysRestService.java [hunk 11]
```diff
@@ -220,8 +224,20 @@ void start() throws Exception {
     ServletContainer sc = new ServletContainer(rc);
     ServletHolder holder = new ServletHolder(sc);
     context.addServlet(holder, "/*");
-
-    jettyServer.start();
+  
+    ServerConnector connector = new ServerConnector(jettyServer);
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/interactivequeries/kafkamusic/MusicPlaysRestService.java]: nit: add `final`

File: src/main/java/io/confluent/examples/streams/interactivequeries/kafkamusic/MusicPlaysRestService.java [hunk 12]
```diff
@@ -220,8 +224,20 @@ void start() throws Exception {
...
+    connector.setHost(hostInfo.host());
+    connector.setPort(hostInfo.port());
+    jettyServer.addConnector(connector);
+  
+    context.start();
+  
+    try {
+      jettyServer.start();
+    } catch (java.net.SocketException exception) {
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/interactivequeries/kafkamusic/MusicPlaysRestService.java]: nit: add `final`

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 13]
```diff
@@ -69,21 +72,44 @@
...
     CLUSTER.createTopic(WordCountInteractiveQueriesExample.TEXT_LINES_TOPIC, 2, 1);
     // The next two topics don't need to be created as they would be auto-created
     // by Kafka Streams, but it just makes the test more reliable if they already exist
     // as creating the topics causes a rebalance which closes the stores etc. So it makes
     // the timing quite difficult...
     CLUSTER.createTopic(WORD_COUNT, 2, 1);
     CLUSTER.createTopic(WINDOWED_WORD_COUNT, 2, 1);
+  
+    // Produce sample data to the input topic before the tests starts.
+    Properties producerConfig = new Properties();
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: nit: add `final`

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 14]
```diff
@@ -139,102 +146,281 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
...
-    final KeyValueBean result = client.target("http://" + hostWithHelloKey.getHost() +
-        ":" + hostWithHelloKey.getPort() +
-        "/state/keyvalue/word-count/hello")
-        .request(MediaType.APPLICATION_JSON_TYPE).get(KeyValueBean.class);
-
-    assertThat(result, equalTo(new KeyValueBean("hello", 2L)));
+  
+    bindHostAndStartRestProxy(port, host);
+  
+    if (proxy != null) {
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Should we fail if `proxy == null` ?

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 15]
```diff
@@ -139,102 +146,281 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
...
+      } catch (Exception ex) {
+        log.error(ex.toString());
+      }
+      count++;
+    }
+  }
+  
+  @Test
+  public void shouldStartRestApiOnAnyValidHost() throws Exception {
+    final int port = randomFreeLocalPort();
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Doesn't using `randomFreeLocalPort()` introduce a race condition? The port might be used by some other process until it's used below?

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 16]
```diff
@@ -139,102 +146,281 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
...
+    kafkaStreams.setStateListener((newState, oldState) -> {
+        if (newState == KafkaStreams.State.RUNNING && oldState == KafkaStreams.State.REBALANCING) {
+          startupLatch.countDown();
+        }
+    });
+    kafkaStreams.start();
+    
+    bindHostAndStartRestProxy(port, host);
+  
+    if (proxy != null) {
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: as above

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 17]
```diff
@@ -139,102 +146,281 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
...
+      assertThat(windowedResult.size(), equalTo(1));
+      final KeyValueBean keyValueBean = windowedResult.get(0);
+      assertTrue(keyValueBean.getKey().startsWith("streams"));
+      assertThat(keyValueBean.getValue(), equalTo(3L));
+    }
+  }
+  
+  @Test(expected = Exception.class)
+  public void shouldThrowExceptionForInvalidHost() throws Exception {
+    final int port = randomFreeLocalPort();
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: as above

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 18]
```diff
@@ -69,21 +72,44 @@
...
+          "streams",
+          "kafka streams");
 
   @Rule
   public final TemporaryFolder temp = new TemporaryFolder();
   private KafkaStreams kafkaStreams;
   private WordCountInteractiveQueriesRestService proxy;
 
+  @Rule
+  public ExpectedException expectedEx = ExpectedException.none();
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: I would prefer not to use this (compare other comment below)

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 19]
```diff
@@ -139,102 +146,281 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
...
+    kafkaStreams.setStateListener((newState, oldState) -> {
+      if (newState == KafkaStreams.State.RUNNING && oldState == KafkaStreams.State.REBALANCING) {
+        startupLatch.countDown();
+      }
+    });
+    
+    kafkaStreams.start();
+    proxy = WordCountInteractiveQueriesExample.startRestProxy(kafkaStreams, port, host);
+    // Binding to same port again will raise BindException.
+    WordCountInteractiveQueriesExample.startRestProxy(kafkaStreams, port, host);
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: We should update this to:
```
try {
    WordCountInteractiveQueriesExample.startRestProxy(kafkaStreams, port, host);
    fail("Should have throw BindException"):
} catch (final BindException expected) {}
```

Otherwise, if a `BindException` is throw before this statement, the test would pass.

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 20]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
     sendPlayEvents(1, songs.get(11), playEventProducer);
 
     playEventProducer.close();
+  }
 
+  private void createStreams(final String host) throws Exception {
+    appServerPort = randomFreeLocalPort();
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: as above

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 21]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+    streams =
+        KafkaMusicExample.createChartsStreams(CLUSTER.bootstrapServers(),
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: nit: why the line break?

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 22]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+                                              appServerPort,
+                                              TestUtils.tempDirectory().getPath(),
+                                              host);
+    int count = 0;
+    int maxTries = 3;
+    while (count <= maxTries) {
+      try {
+        // Starts the Rest Service on the provided host:port
+        restProxy = KafkaMusicExample.startRestProxy(streams, new HostInfo(host, appServerPort));
+      } catch (Exception ex) {
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: nit: add `final`

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 23]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+  private void createStreams(final String host) throws Exception {
+    appServerPort = randomFreeLocalPort();
+    streams =
+        KafkaMusicExample.createChartsStreams(CLUSTER.bootstrapServers(),
+                                              CLUSTER.schemaRegistryUrl(),
+                                              appServerPort,
+                                              TestUtils.tempDirectory().getPath(),
+                                              host);
+    int count = 0;
+    int maxTries = 3;
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: nit: add `final`

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 24]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+      }
+      count++;
+    }
+  }
 
+  @After
+  public void shutdown() throws Exception {
+    try {
+      restProxy.stop();
+    } catch (Exception e) {
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: nit: add `final`

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 25]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+    streams.close();
+  }
+
+  @Test
+  public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Exception {
+    final String host = "localhost";
+    createStreams(host);
     streams.start();
 
+  if (restProxy != null) {
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: as above

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 26]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+                              songCountPlayBean(songs.get(2), 4L)
+                              )
+                );
+
+    }
+  }
+
+  @Test
+  public void shouldDemonstrateInteractiveQueriesOnAnyValidHost() throws Exception {
+  final String host = "127.10.10.10";
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: fix indention

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 27]
```diff
@@ -210,17 +148,108 @@ public void shouldCreateChartsAndAccessThemViaInteractiveQueries() throws Except
...
+  createStreams(host);
+  streams.start();
+
+  if (restProxy != null) {
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: as above

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 28]
```diff
@@ -299,7 +329,7 @@ private void verifyChart(final String url,
     assertThat(chart, is(expectedChart));
   }
 
-  private void sendPlayEvents(final int count, final Song song,
+  private static void sendPlayEvents(final int count, final Song song,
                               final KafkaProducer<String, PlayEvent> producer) {
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: Nit: fix indention and move `final Song song` to it's own line

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java (hunk 15 above)
Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Yes @mjsax, it may introduce. And, Therefore, the race condition has already been taken care of using a try/catch block while starting the server.
You can find it here:
WordCountInteractiveQueriesRestService#start(final int port)

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java (hunk 14 above)
Comment (nehabhardwaj01) [reply src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: Yes, we can. Will do the required change @mjsax

**Comment by nehabhardwaj01** [comment]
Hi @mjsax,
Thanks for reviewing the PR, I've added the suggested changes. Will be great if you can review them once.

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 29]
```diff
@@ -244,6 +245,8 @@ public void shouldDemonstrateInteractiveQueries() throws Exception {
       final KeyValueBean keyValueBean = windowedResult.get(0);
       assertTrue(keyValueBean.getKey().startsWith("streams"));
       assertThat(keyValueBean.getValue(), equalTo(3L));
+    } else {
+        fail("Should fail demonstrating InteractiveQueries as the Rest Service failed to start.");
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: nit: fix indention -- should be 2 spaces

File: src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java [hunk 30]
```diff
@@ -377,6 +380,8 @@ public void shouldStartRestApiOnAnyValidHost() throws Exception {
       final KeyValueBean keyValueBean = windowedResult.get(0);
       assertTrue(keyValueBean.getKey().startsWith("streams"));
       assertThat(keyValueBean.getValue(), equalTo(3L));
+    } else {
+        fail("Should fail demonstrating InteractiveQueries on any valid host as the Rest Service failed to start.");
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/WordCountInteractiveQueriesExampleTest.java]: nit: fix indention -- should be 2 spaces

File: src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java [hunk 31]
```diff
@@ -152,20 +153,19 @@ public static void createTopicsAndProduceDataToInputTopics() throws Exception {
 
   private void createStreams(final String host) throws Exception {
     appServerPort = randomFreeLocalPort();
-    streams =
-        KafkaMusicExample.createChartsStreams(CLUSTER.bootstrapServers(),
+    streams = KafkaMusicExample.createChartsStreams(CLUSTER.bootstrapServers(),
                                               CLUSTER.schemaRegistryUrl(),
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/interactivequeries/kafkamusic/KafkaMusicExampleTest.java]: nit: fix indention -- parameters should align (also lines below)

**Comment by nehabhardwaj01** [comment]
Hi @mjsax, I've checked in the latest code. Can you please verify the addressed changes.

**Comment by mjsax** [comment]
Thanks a lot for the PR @nehabhardwaj01!
